prow/cmd/checkconfig/main.go

```diff
@@ -212,6 +236,91 @@ func validateURLs(c config.ProwConfig) error {
 	return errorutil.NewAggregate(validationErrs...)
 }
 
+func validateUnknownFields(cfg interface{}, cfgBytes []byte, filePath string) error {
+	var obj interface{}
+	err := yaml.Unmarshal(cfgBytes, &obj)
+	if err != nil {
+		return fmt.Errorf("error while unmarshaling yaml: %v", err)
+	}
+	unknownFields := checkUnknownFields("", obj, reflect.ValueOf(cfg))
+	if len(unknownFields) > 0 {
+		sort.Strings(unknownFields)
+		return fmt.Errorf("unknown fields present in %s: %v", filePath, strings.Join(unknownFields, ", "))
+	}
+	return nil
+}
+
+func checkUnknownFields(keyPref string, obj interface{}, cfg reflect.Value) []string {
+	var uf []string
+	switch concreteVal := obj.(type) {
+	case map[string]interface{}:
+		// Iterate over map and check every value
+		for key, val := range concreteVal {
+			fullKey := fmt.Sprintf("%s.%s", keyPref, key)
+			subCfg := getSubCfg(key, cfg)
+			if !subCfg.IsValid() {
+				// Append fullKey without leading "."
+				uf = append(uf, fullKey[1:])
+			} else {
+				subUf := checkUnknownFields(fullKey, val, subCfg)
+				uf = append(uf, subUf...)
+			}
+		}
+	case []interface{}:
+		for i, val := range concreteVal {
+			fullKey := fmt.Sprintf("%s[%v]", keyPref, i)
+			subCfg := cfg.Index(i)
+			uf = append(uf, checkUnknownFields(fullKey, val, subCfg)...)
+		}
+	}
+	return uf
+}
+
+func getSubCfg(key string, cfg reflect.Value) reflect.Value {
+	cfgElem := cfg
+	if cfg.Kind() == reflect.Interface || cfg.Kind() == reflect.Ptr {
+		cfgElem = cfg.Elem()
+	}
+	switch cfgElem.Kind() {
+	case reflect.Map:
+		iter := cfgElem.MapRange()
+		for iter.Next() {
+			k := iter.Key().Interface().(string)
```

Comment:
What if this cast fails?

Reply:
Good point. I thought about it and haven't found what would be the example of such scenario (map with integer keys?).
But you're right, even if I can't think of scenarios it have to be handled and not fail. Fixed.